core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -650,14 +650,24 @@ private IndexResult innerIndex(Index index) throws IOException {
                 }
                 updatedVersion = index.versionType().updateVersion(currentVersion, expectedVersion);
                 index.parsedDoc().version().setLongValue(updatedVersion);
+
+                // Update the document's sequence number and primary term, the
+                // sequence number here is derived here from either the sequence
+                // number service if this is on the primary, or the existing
+                // document's sequence number if this is on the replica. The
+                // primary term here has already been set, see
+                // IndexShard.prepareIndex where the Engine.Index operation is
+                // created
+                index.parsedDoc().updateSeqID(seqNo, index.primaryTerm());
```

Comment:
I'm totally open to suggestions otherwise though, @jasontedor you're about to use this PR, do you have a preference?